add logic to check for first responder job type

phillyCHEEEEEZ opened the pull request:
Dispatch events being called with `job = {"FirstResponder"}` were not showing up in the call list on the dashboard and map pages of the MDT.  This PR adds additional logic to the job checks which compares the data passed to "FirstResponder" so that they may be properly displayed.

phillyCHEEEEEZ committed:
add logic to check for first responder job type

xFutte commented on ui/app.js:
```diff
@@ -4370,7 +4370,7 @@ $(document).ready(() => {
       const value = eventData.data;
       DispatchMAP(value);
-      if (value && value.job.includes(playerJob)) {
+      if (value && (value.job.includes(playerJob) || value.job == "FirstResponder")) {
```
I'm fairly sure this doesn't work. `value.job` looks to me being an array, meaning that `value.job == "FirstResponder"` won't work. Basically you are asking if an array is a string value. Same goes for the other if statement.

phillyCHEEEEEZ replied:
I actually thought this was the case when I originally look at it as well, but during my tests I found that logging the value of `value.job` when "FirstResponder" is passed appeared to log it as a string which is why I wrote it this way.  And after making this change on my local it caused the call list to start getting populated with calls where "FirstResponder" is passed whereas previously none were being displayed.

You do bring up a good point that I had overlooked, however.  Matching against a single string would only be true if that was the only value that was passed.  There may be cases where "FirstResponder" isn't the only piece of data passed to the job field.  I've made an update to adjust the logic to utilize the `includes()` method to check if that string exists anywhere in `value.job`.  That should avoid the aforementioned issues while still properly displaying dispatch calls in the call lists.

phillyCHEEEEEZ committed:
improve logic to check for first responder job

previous iteration only checked for a single string to match against.  this is not ideal any data passed that is more than just that single string might fail.  checking the entire `value.job` field to see if it includes the text we are trying to match against prevents this issue.

BackSH00TER commented on ui/app.js:
```diff
@@ -4370,7 +4370,7 @@ $(document).ready(() => {
       const value = eventData.data;
       DispatchMAP(value);
-      if (value && value.job.includes(playerJob)) {
+      if (value && (value.job.includes(playerJob) || value.job.includes("FirstResponder"))) {
```
This seems like a very specific change, that might be better suited to add into the config or on your specific fork, rather than added into the framework. The config has sections for Config.PoliceJobs,  Config.AmbulanceJobs, Config.DojJobs

I'm curious how many folks actually have a job "FirstResponder" or if this is more so a specific job type you have on your server? IE: Most servers have "police" and "ambulance"

We should aim to support the main job types in this framework rather than add specific job types that may not be used by the majority of consumers.

phillyCHEEEEEZ replied:
Allow me to address a few of your comments.

> This seems like a very specific change, that might be better suited to add into the config or on your specific fork, rather than added into the framework.

This isn't something that I built custom for my own fork/server nor is it even a job that I have configured on my end.  This is functionality that ps-dispatch comes with out of the box.  More info on this below.

> The config has sections for Config.PoliceJobs, Config.AmbulanceJobs, Config.DojJobs

This is true, but the out of the box functionality I alluded to above doesn't utilize those tables displaying calls in the MDT.  It would require the player's job to be "FirstResponder" which, as you have already said, is non-conventional as most servers use police/ambulance/etc.

I encourage you to verify this issue yourself by doing the following:

1. Launch a server with ps-mdt and ps-dispatch
2. Set yourself to one of the jobs listed in the config tables you mentioned above
3. Trigger an officer/EMS down alert
4. Notice that the dispatch alert gets sent normally but the MDT call lists do not get populated with said call

> I'm curious how many folks actually have a job "FirstResponder" or if this is more so a specific job type you have on your server? IE: Most servers have "police" and "ambulance"

To expand on the out of the box functionality I've been talking about... this "FirstResponder" value is something that ps-dispatch is configured by default to pass as the only piece of data to several alerts so every single user will experience this issue out of the box.  See below:

- https://github.com/Project-Sloth/ps-dispatch/blob/e31944a7059ddaf4e95722885dd5a6a252c51898/client/cl_events.lua#L549
- https://github.com/Project-Sloth/ps-dispatch/blob/e31944a7059ddaf4e95722885dd5a6a252c51898/client/cl_events.lua#L581

Those two dispatch alerts will not show up in the MDT call lists of any player unless the same modifications shown here are made on their end or add additional jobs to the data that gets passed, which some people may not realize they need to do.  This will leave them wondering why when a dispatch alert gets sent for officer/EMS down it never shows up in the call list resulting in requests for support for something that they think is broken, but is *technically* working. Another possible solution was to edit the ps-dispatch functions to pass additional jobs, but I opted to go this route to address the issue of a user wanting to pass just a single "FirstResponder" value, especially considering that is how it is currently configured to function.  I see these aforementioned issues as problematic and this proposed solution is quite simple and addresses all of them.

MonkeyWhisper replied:
Both an oversight on there, we will accept a PR correcting that so it sends to all jobs. 

I'm not sure we had a FirstResponders job, that's just ambulance
https://github.com/Project-Sloth/ps-dispatch/blob/2c4f9fdfde24cfd327c9c47ba0ec35a50ef61028/config.lua#L70

@DonHulieo was this part of your changes by chance? I doubt it

DonHulieo replied:
> @DonHulieo was this part of your changes by chance? I doubt it

Nah mate, I haven’t made any PR’s to MDT. I haven’t implemented my method for checking police jobs in the MDT yet, and calling FirstResponder like that (because as you said it’s not a ‘job’) will just return nil in a traditional job check without my additional logic.

MonkeyWhisper replied:
> > @DonHulieo was this part of your changes by chance? I doubt it
> 
> Nah mate, I haven’t made any PR’s to MDT. I haven’t implemented my method for checking police jobs in the MDT yet, and calling FirstResponder like that (because as you said it’s not a ‘job’) will just return nil in a traditional job check without my additional logic.

Yeah, I figured as much. That was part of the ps-dispatch, since you have been working on the dispatch, you think you can work on this for the dispatch? We'll then go ahead and adjust things on ps-mdt

DonHulieo replied:
> figured as much. That was part of the ps-dispatch, since you have been working

If I have edit permission for this PR, I can see if I can implement my changes on here, or should I just create my own PR?

BackSH00TER replied:
@phillyCHEEEEEZ Thanks for the details, wasn't aware that was part of dispatch. Looks like we've surfaced the issue tho and got the right conversation going.

MonkeyWhisper replied:
> 

Feel free to send your own PR and we will get @phillyCHEEEEEZ to edit this one as well so they can get credit for this 

Thanks everyone
